gordon_janitor_gcp/auth.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Module to create a client interacting with Google Cloud authentication.
+
+An instantiated client is needed for interacting with any of the Google
+APIs via the ``gordon_janitor_gcp.http_client``.
+
+Only service account (JSON Web Tokens/JWT) authentication is currently
+supported. To setup a service account, follow `Google's docs <https://
+cloud.google.com/iam/docs/creating-managing-service-account-keys>`_.
+
+To use:
+
+.. code-block:: pycon
+
+    >>> import asyncio
+    >>> from google_janitor_gcp import auth
+    >>> loop = asyncio.get_event_loop()
+    >>> keyfile = '/path/to/service_account_keyfile.json'
+    >>> creds = auth.GoogleAuthClient(keyfile=keyfile)
+    >>> creds.token is None
+    True
+    >>> loop.run_until_complete(creds.refresh_token())
+    >>> creds.token
+    'c0ffe3'
+
+"""
+
+import asyncio
+import json
+import logging
+import urllib.parse
+
+import aiohttp
+from google.oauth2 import _client
+from google.oauth2 import service_account
+
+from gordon_janitor_gcp import exceptions
+
+
+REQ_LOG_FMT = 'Request: "{method} {url}"'
+RESP_LOG_FMT = 'Response: "{method} {url}" {status} {reason}'
+
+
+class GoogleAuthClient:
+    """Async client to authenticate against Google Cloud APIs.
+
+    Attributes:
+        SCOPE_TMPL_URL (str): template URL for Google auth scopes.
+        DEFAULT_SCOPE (str): default scope if not provided.
+        JWT_GRANT_TYPE (str): grant type header value when
+            requesting/refreshing an access token.
+
+    Args:
+        keyfile (str): path to service account (SA) keyfile.
+        scopes (list): (optional) scopes with which to authorize the SA.
+            Default is ``'cloud-platform'``.
+        session (aiohttp.ClientSession): (optional) ``aiohttp`` HTTP
+            session to use for sending requests.
+        loop: (optional) asyncio event loop to use for HTTP requests.
+            NOTE: if ``session`` is given, then ``loop`` will be ignored.
+            Otherwise, ``loop`` will be used to create a session, if
+            provided.
+
+    """
+
+    SCOPE_TMPL_URL = 'https://www.googleapis.com/auth/{scope}'
+    DEFAULT_SCOPE = 'cloud-platform'
+    JWT_GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:jwt-bearer'
```

Comment:
Should these be (semi) private?

Reply:
I don't have a strong opinion on this. I always understood class attributes as already sort of private because they're specific to a class and shared by all the objects created by that class, something you don't want to go and change.

Reply:
fair enough - keeping it as is.